src/atlas/array/helpers/ArrayForEach.h

```diff
@@ -114,28 +129,50 @@ constexpr auto argPadding() {
   }
 }
 
-template <typename... SlicerArgs, template <typename, int> typename View,
-          typename Value, int Rank, typename... ArrayViews>
+template<int Rank>
+struct GetRankImpl {
+  constexpr static int rank = Rank;
+};
+
+template<typename>
+struct GetRank;
+
+template<template <typename, int> typename View, typename Value, int Rank>
+struct GetRank<View<Value, Rank>> : GetRankImpl<Rank> {};
+template<template <typename, int> typename View, typename Value, int Rank>
+struct GetRank<View<Value, Rank>&> : GetRankImpl<Rank> {};
+template<template <typename, int> typename View, typename Value, int Rank>
+struct GetRank<const View<Value, Rank>&> : GetRankImpl<Rank> {};
+template<template <typename, int> typename View, typename Value, int Rank>
+struct GetRank<View<Value, Rank>&&> : GetRankImpl<Rank> {};
+template<template <typename, int> typename View, typename Value, int Rank>
+struct GetRank<const View<Value, Rank>&&> : GetRankImpl<Rank> {};
+
+template <size_t ViewIdx = 0, typename... SlicerArgs, typename ArrayViewTuple>
 auto makeSlices(const std::tuple<SlicerArgs...>& slicerArgs,
-                View<Value, Rank>& arrayView, ArrayViews&... arrayViews) {
+                ArrayViewTuple&& arrayViews) {
 
-  // "Lambdafy" slicer apply method to work with std::apply.
-  const auto slicer = [&arrayView](const auto&... args) {
-    return std::make_tuple(arrayView.slice(args...));
-  };
+  using ViewTupleType = std::decay_t<ArrayViewTuple>;
 
-  // Fill out the remaining slicerArgs with Range::all().
-  constexpr auto Dim = sizeof...(SlicerArgs);
-  const auto paddedArgs = std::tuple_cat(slicerArgs, argPadding<Rank - Dim>());
+  if constexpr(ViewIdx < std::tuple_size_v<ViewTupleType>) {
 
-  if constexpr(sizeof...(ArrayViews) > 0) {
+    auto&& view = std::get<ViewIdx>(arrayViews);
+    using View = std::tuple_element_t<ViewIdx, ViewTupleType>;
+
+    constexpr auto Dim = sizeof...(SlicerArgs);
+    constexpr auto Rank = GetRank<View>::rank;
```

Comment:
Wouldn't this work: 
```c++
constexpr auto Rank = View::rank();
```
or maybe even:
```c++
constexpr auto Rank = view.rank();
```
In that latter case you don't need to define `View` type

Reply:
The first one works a treat.